pkg/controller/client.go

```diff
@@ -0,0 +1,194 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc"
+
+	"kmesh.net/kmesh/pkg/controller/config"
+	"kmesh.net/kmesh/pkg/controller/envoy"
+	"kmesh.net/kmesh/pkg/controller/workload"
+	"kmesh.net/kmesh/pkg/nets"
+)
+
+const (
+	RandTimeSed = 1000
+)
+
+type XdsClient struct {
+	ctx            context.Context
+	cancel         context.CancelFunc
+	grpcConn       *grpc.ClientConn
+	client         service_discovery_v3.AggregatedDiscoveryServiceClient
+	adsStream      *envoy.AdsStream
+	workloadStream *workload.WorkloadStream
+	xdsConfig      *config.XdsConfig
+}
+
+func NewXdsClient() *XdsClient {
+	client := &XdsClient{
+		xdsConfig: config.GetConfig(),
+		adsStream: &envoy.AdsStream{
+			Event: envoy.NewServiceEvent(),
+		},
+		workloadStream: &workload.WorkloadStream{
+			Event: workload.NewServiceEvent(),
+		},
+	}
+
+	client.ctx, client.cancel = context.WithCancel(context.Background())
+
+	return client
+}
+
+func (c *XdsClient) createStreamClient() error {
+	var err error
+
+	if c.grpcConn, err = nets.GrpcConnect(c.xdsConfig.DiscoveryAddress); err != nil {
+		return fmt.Errorf("grpc connect failed: %s", err)
+	}
+
+	c.client = service_discovery_v3.NewAggregatedDiscoveryServiceClient(c.grpcConn)
+
+	if bpfConfig.EnableKmesh {
+		if err = c.adsStream.AdsCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create ads stream failed, %s", err)
+		}
+	} else if bpfConfig.EnableKmeshWorkload {
+		if err = c.workloadStream.WorklaodCreateStream(c.client, c.ctx); err != nil {
+			return fmt.Errorf("create workload stream failed, %s", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *XdsClient) recoverConnection() error {
+	var (
+		err      error
+		interval = time.Second
+	)
+
+	c.closeStreamClient()
+	for count := 0; count < nets.MaxRetryCount; count++ {
+		if err = c.createStreamClient(); err == nil {
+			return nil
+		}
+
+		log.Errorf("grpc connect failed, %s", err)
+		c.closeStreamClient()
+		time.Sleep(interval + nets.CalculateRandTime(RandTimeSed))
+		interval = nets.CalculateInterval(interval)
+	}
+
+	return fmt.Errorf("retry %d times", nets.MaxRetryCount)
+}
+
+func (c *XdsClient) clientProcess(ctx context.Context) {
+	var (
+		err            error
+		reconnect      = false
+		isConnectAgain = new(bool)
+	)
+
+	*isConnectAgain = false
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if reconnect {
+				log.Warnf("reconnect due to %s", err)
+				if err = c.recoverConnection(); err != nil {
+					log.Errorf("recover connection failed, %s", err)
+					return
+				}
+				reconnect = false
+				if bpfConfig.EnableKmeshWorkload {
+					*isConnectAgain = true
```

Comment:
Why is this variable only used in workload mode? AFAIK, the behavior of ads mode and workload mode is similiar.

In addition, it's  unnecessary to pass `isConnectAgain` as pointer, looks weird?

Reply:
workload currently uses delta subscription, but when the connection is reconnected, it will push the workload info in full again, which may cause duplicate info storage. Passing through a pointer is because the flag will be reset in the WorkloadStreamProcess function

Reply:
I donot think skip procesing after reconnect is right. Because during the period, the workloads can be updated

Reply:
fixed